proposal on replacing OKLog with Loki (#199)
kyma-project/community: sig-and-wg/sig-core/proposals/loki-as-bundled-logging-solution.md

# Loki as bundled logging solution

Created on 2019-01-28 by Andreas Thaler (@a-thaler).

## Status

Proposed on 2019-01-28

## Motivation

> Kyma should provide a "batteries-included"-like solution for logging. Such a solution must be very lightweight as it must run in-cluster and should work for local development. The current approach is to use [OK Log](https://github.com/oklog/oklog), which is fulfilling the main criteria, but unfortunately its development has been discontinued and it also lacks a lot of user experience features.

Reviewer: ```suggestion
Kyma should provide a "batteries-included" solution for logging. Such a solution must be very lightweight as it must run in-cluster and should work for a local development. The current approach is to use [OK Log](https://github.com/oklog/oklog), which is fulfilling the main criteria, but unfortunately its development has been discontinued and it also lacks many user experience features.
```

## Goal
Replace OKLog by a solution which is:
- leightweight in resource consumption (in-cluster local development support)
- easy to operate even when scaling up the cluster
- allow to query recent logs for a pod in a user-friendly way
- maintained and further developed by a community

## Proposal
Replace OKLog with [Loki](https://github.com/grafana/loki) by:

> - Adjust the logging helm chart to deploy the Loki service and the promtail daemonSet instead of OKLog service and logspout daemonSetpromtail

Author: The promtail component gets deployed as a DaemonSet, which is a type of a deployment. So it is a promtail DaemonSet

- Enable the experimental feature of grafana to enable the new exporer and integration with Loki as data source

## Details
Loki is a new project by Grafana Labs and aims to provide a similar solution for logging, like Prometheus is providing for monitoring.

> It uses a similar approach like OKLog by explicitly not doing a full text indexing, but just an indexing on metadata level. Here, it is leveraging the existing labels of Prometheus and Kubernetes and it integrates with Grafana out-of-the-box (experimental feature for now).

Reviewer: ```suggestion
It uses a similar approach as the OK Log by indexing on the metadata level instead of performing full-text indexing. Additionally, it is leveraging the existing labels of Prometheus and Kubernetes and it integrates with Grafana out-of-the-box (experimental feature for now).
```

> It uses a similar approach as the OK Log by indexing on the metadata level instead of performing full-text indexing. Additionally, it is leveraging the existing labels of Prometheus and Kubernetes and it integrates with Grafana out-of-the-box (experimental feature for now).

Reviewer: ```suggestion
It uses a similar approach as the OK Log by indexing on the metadata level instead of performing full-text indexing. Additionally, it is leveraging the existing labels of Prometheus and Kubernetes and it integrates with Grafana out-of-the-box (an experimental feature for now).
```

> Besides OKLog and Loki there was no other alternative covering the described requirements.

Reviewer: ```suggestion
Apart from the OK Log and Loki there was no other alternative covering the described requirements.
```

> OKLog should be replaced with Loki as

Reviewer: ```suggestion
OK Log should be replaced with Loki because:
```

> - it has an active development community

Reviewer: ```suggestion
- It has an active development community.
```

> - is leveraging the existing labels and with that provides a much better search experience

Reviewer: ```suggestion
- It is leveraging the existing labels, thus providing a much better search experience.
```

> - It is leveraging the existing labels, thus providing a much better search experience.

Reviewer: ```suggestion
- It leverages the existing labels, thus providing a much better search experience.
```

> - it provides the better Graphical user interface by being integrated to Grafana

Reviewer: ```suggestion
- It provides an improved graphical user interface as a result of integration with Grafana. 
```
